libraries/apollo-tooling/src/main/kotlin/com/apollographql/apollo3/tooling/SchemaUploader.kt

```diff
@@ -67,27 +66,30 @@ object SchemaUploader {
         .httpHeaders(headers.map { HttpHeader(it.key, it.value) })
     val response = runBlocking { call.execute() }
     val data = response.data
-    if (data == null) {
-      when (val e = response.exception!!) {
-        is ApolloHttpException -> {
-          val body = e.body?.use { it.readUtf8() } ?: ""
-          throw Exception("Cannot upload schema: (code: ${e.statusCode})\n$body", e)
-        }
-
-        is ApolloGraphQLException -> {
-          throw Exception("Cannot upload schema: ${e.errors.joinToString { it.message }}")
-        }
+    if (data != null) {
+      if (data.graph == null) {
+        throw Exception("Cannot retrieve graph '$graphID': ${response.errors?.joinToString { it.message }}")
+      }
+      val code = data.graph.uploadSchema.code
+      val message = data.graph.uploadSchema.message
+      val success = data.graph.uploadSchema.success
+      if (!success) {
+        throw Exception("Cannot upload schema (code: $code): $message")
+      }
+      return
+    }
+    when (val e = response.exception) {
+      is ApolloHttpException -> {
+        val body = e.body?.use { it.readUtf8() } ?: ""
+        throw Exception("Cannot upload schema: (code: ${e.statusCode})\n$body", e)
+      }
 
-        else -> {
-          throw Exception("Cannot upload schema: ${e.message}", e)
-        }
+      null -> {
+        throw Exception("Cannot upload schema: ${response.errors?.joinToString { it.message }}")
       }
-    } else {
-      val code = data.graph?.uploadSchema?.code
-      val message = data.graph?.uploadSchema?.message
-      val success = data.graph?.uploadSchema?.success
-      check(success == true) {
-        "Cannot upload schema (code: $code): $message"
+
+      else -> {
+        throw Exception("Cannot upload schema: ${e.message}", e)
```

Comment:
Should we make `ApolloGraphQLException.message` receive the GraphQL error's message? Currently it's a generic `"GraphQL error"`. (We land on this `else` case if `graph` isn't null but `uploadSchema` is.)

Reply:
Yes good idea 👍